src/pipdeptree/_models/package.py

```diff
@@ -118,31 +110,69 @@ class DistPackage(Package):
 
     """
 
-    def __init__(self, obj: DistInfoDistribution, req: ReqPackage | None = None) -> None:
-        super().__init__(obj)
+    def __init__(self, obj: Distribution, req: ReqPackage | None = None) -> None:
+        super().__init__(obj.metadata["Name"])
+        self._obj = obj
         self.req = req
-        self._project_name = ""
 
     def requires(self) -> list[Requirement]:
-        return self._obj.requires()  # type: ignore[no-untyped-call,no-any-return]
+        req_list = []
+        req_name_list = []
+        if self._obj.requires:
+            for r in self._obj.requires:
+                req = Requirement(r)
+                is_extra_req = req.marker and contains_extra(str(req.marker))
+                if not is_extra_req and req.name not in req_name_list:
+                    req_list.append(req)
+                    req_name_list.append(req.name)
+        return req_list
 
     @property
-    def project_name(self) -> str:
-        if not self._project_name:
-            try:
-                self._project_name = metadata(self.key)["name"]
-            except (PackageNotFoundError, KeyError):
-                self._project_name = self._obj.project_name
-        return self._project_name
+    def editable(self) -> bool:
+        return self.direct_url_dict["editable"]
+
+    @property
+    def direct_url(self) -> str:
+        return self.direct_url_dict["url"]
```

Comment:
pip expects this to return a `DirectUrl` object, but here we return a `str`. Because of this, we will crash when we reach this [line](https://github.com/pypa/pip/blob/4f3a1847c98323b0492a4b6cf125f57684f49344/src/pip/_internal/utils/direct_url_helpers.py#L11C1).

Reply:
I learned the implementation of direct_url in pip again and re-implemented this part. Now it has almost the same logic  as it is in [pip](https://github.com/pypa/pip/blob/f5e4ee104e7b171a7cfb2843c9c602abf7a4e346/src/pip/_internal/metadata/base.py#L289).

Reply:
At first, we thought that making compatible to frozen API only needed simple construction, but it was not the case, so it was better for us to fully implement this part ourselves.

Reply:
Done